drivers/FlashIAP.cpp

```diff
@@ -0,0 +1,129 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2017 ARM Limited
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "FlashIAP.h"
+#include "mbed_assert.h"
+
+#ifdef DEVICE_FLASH
+
+namespace mbed {
+
+FlashIAP::FlashIAP()
+{
+
+}
+
+FlashIAP::~FlashIAP()
+{
+
+}
+
+int FlashIAP::init()
```

Comment:
I think typdefs help document the intent of different parts of the function declaration:
``` cpp
flash_error_t FlashIAP::read(void *buffer, flash_addr_t addr, flash_size_t size);
```

However I don't think the error return should be an enum. It allows compilation warnings, but prevents cross module errors and the strict casting rules of c++ limit compatibility with simple ints (which I don't think is unreasonable for users to use). Keeping error types a typedef of int allows -1 as a "miscellaneous" error and gradual adoption of more useful error codes.

Just my 2c, the new classes do diverge here, so a different perspective may be valuable.